feature: add new lint `pub_underscore_fields`

ParkMyCar opened the pull request:
fixes: #10282 

This PR introduces a new lint `pub_underscore_fields` that lints when a user has marked a field of a struct as public, but also prefixed it with an underscore (`_`). This is something users should avoid because the two ideas are contradictory. Prefixing a field with an `_` is inferred as the field being unused, but making a field public infers that it will be used.

- \[x] Followed [lint naming conventions][lint_naming]
  - I believe I followed the naming conventions, more than happy to update the naming if I did not :)
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[x] `cargo test` passes locally
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`

---

changelog: new lint: [`pub_underscore_fields`]
[#10283](https://github.com/rust-lang/rust-clippy/pull/10283)
<!-- changelog_checked -->

ParkMyCar commented:
Hey @flip1995, if you wouldn't mind taking a look at this when you get the chance, it would be much appreciated! No rush though :)

bors commented:
:umbrella: The latest upstream changes (presumably #10481) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

blyxyas commented:
@ParkMyCar Philipp stopped being on the reviewer rotation a few months ago, are you still interested in continuing with this development?

r? @blyxyas

ParkMyCar commented:
> @ParkMyCar Philipp stopped being on the reviewer rotation a few months ago, are you still interested in continuing with this development? ฅ^•ﻌ•^ฅ
> 
> r? @blyxyas

For sure! I just rebased off of master and updated the lint version, whenever you have time let me know if there is any other feedback :)

blyxyas commented on clippy_lints/src/pub_underscore_fields.rs:
````diff
@@ -0,0 +1,57 @@
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// struct FileHandle {
+    ///     _descriptor: usize,
+    /// }
+    /// ```
````
The other alternative is also needed :cat2:

```suggestion
    /// ```rust
    /// struct FileHandle {
    ///     _descriptor: usize,
    /// }
    /// // OR 
    /// struct FileHandle {
    ///     pub descriptor: usize,
    /// }
    /// ```
```

ParkMyCar replied:
Ah ha, so right, updated!

blyxyas reviewed (changes requested):
Very good for a first review! Just a couple minor changes and I think this is it :heart: |ΦωΦ|

ParkMyCar commented on clippy_lints/src/pub_underscore_fields.rs:
```diff
@@ -0,0 +1,57 @@
+            return;
+        };
+
+        let msg = "field marked as public but also inferred as unused because it's prefixed with `_`";
```
Inlined it is :)

ParkMyCar commented on clippy_lints/src/pub_underscore_fields.rs:
```diff
@@ -0,0 +1,57 @@
+        };
+
+        let msg = "field marked as public but also inferred as unused because it's prefixed with `_`";
+        for field in st.fields().iter() {
```
Done!

ParkMyCar commented:
> It would be good if this ignored `PhantomData` fields (it's common to underscore prefix them) and fields that are `#[doc(hidden)]`

Thanks for the feedback @Alexendoo! I'm not sure I agree though. AFAIU prefixing a field name with an underscore generally indicates that field should be ignored, if some part of your code needs to use the `PhantomData` marker, then it should not be prefixed, same with `#[doc_hidden]`. I'd be more than happy to make this change, but it's not totally clear to me why these would need to be special cased, could you explain more or provide an example? Thanks :)

blyxyas commented on clippy_lints/src/pub_underscore_fields.rs:
````diff
@@ -20,9 +20,15 @@ declare_clippy_lint! {
     /// Use instead:
     /// ```rust
-    /// struct FileHandle {
+    /// struct FileHandle_Foo {
````
Mixing PascalCase with underscores hehe :cat2:
For future reference, you can use [attributes](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#attributes) like `should_panic` on your code blocks so that these name collisions don't matter.

Alexendoo replied:
You can make it two separate code blocks
`````md
```rust
struct FileHandle { .. }
```
or
```rust
struct FileHandle { .. }
```
`````

ParkMyCar replied:
🤦‍♂️ yep totally missed this, updated it to be two code blocks!

Alexendoo commented:
`doc(hidden)` + underscore prefix is used to drive home the "please don't use this" message, that the field is only public for some implementation reason such as being used in a macro expansion

e.g. https://docs.rs/enumset/1.1.2/src/enumset/set.rs.html#125-130

ParkMyCar commented:
@blyxyas @Alexendoo thanks for the examples! I'm thoroughly convinced, I updated the lint to be a LatePass so we can get type info. Then skip fields that are `#[doc(hidden)] or use `PhantomData`.

I struggled a bit to get the visibility settings right, if there's an easier to detect the range of "pub" visibilities we want to guard against, please let me know!

ParkMyCar commented:
Whoops, just realized I forgot to add tests for the new `#[doc(hidden)]` and `PhantomData` checks, let me do that!

blyxyas commented on clippy_lints/src/pub_underscore_fields.rs:
```diff
@@ -37,21 +43,36 @@ declare_clippy_lint! {
+impl<'tcx> LateLintPass<'tcx> for PubUnderscoreFields {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'_>) {
+        // This lint only pertains to structs.
+        let ItemKind::Struct(var, _) = &item.kind else {
```
`var` kinda implies "variable", I think the full name `variant_data` would be more illustrative (even more now, that IDEs autocomplete variable names)

blyxyas commented on clippy_lints/src/pub_underscore_fields.rs:
```diff
@@ -1,6 +1,10 @@
-use rustc_ast::ast::{Item, ItemKind, VisibilityKind};
-use rustc_lint::{EarlyContext, EarlyLintPass};
+use clippy_utils::is_path_lang_item;
+// use rustc_ast::ast::{Item, ItemKind, VisibilityKind};
```
```suggestion
```

blyxyas commented on clippy_lints/src/pub_underscore_fields.rs:
````diff
@@ -20,13 +24,15 @@ declare_clippy_lint! {
-    /// 
-    /// struct FileHandle_Bar {
+    ///
+    /// ```
````
I'm not sure if the markdown page will get this, so just in case ミ๏ｖ๏彡

```suggestion
    /// ```rust
```

blyxyas commented on clippy_lints/src/pub_underscore_fields.rs:
```diff
@@ -37,21 +43,36 @@ declare_clippy_lint! {
+
+        for field in var.fields() {
+            // Only pertains to fields that start with an underscore, and are visible publically.
+            if field.ident.as_str().starts_with('_') && is_visible(&field)
```
```suggestion
            if field.ident.as_str().starts_with('_') && is_visible(field)
```

blyxyas reviewed (approved):
Just some polishing and I think this is ready! Thanks for the patience! :heart:  (=^･ω･^=)

Alexendoo commented on clippy_lints/src/pub_underscore_fields.rs:
```diff
@@ -0,0 +1,83 @@
+            return;
+        };
+
+        let is_visible = |field: &FieldDef<'_>| {
+            let parent = cx.tcx.parent_module_from_def_id(field.def_id);
+            let grandparent = cx.tcx.parent_module_from_def_id(parent.into());
+            let visibility = cx.tcx.visibility(field.def_id);
+
+            let case_1 = parent == grandparent && !field.vis_span.is_empty();
+            let case_2 = visibility != Visibility::Restricted(parent.to_def_id());
+
+            case_1 || case_2
+        };
```
Can try `cx.effective_visibilities.is_exported(field.def_id))` for this

ParkMyCar replied:
I tried that, unfortunately it didn't work :/

Alexendoo replied:
I think the issue is the test file, none of the fields in it are actually publicly reachable. You'd have to move the definitions outside of `fn main() {}` and make the `inner`/`inner2` modules public

blyxyas replied:
`is_exported` doesn't lint for `pub(...)`, and it seems like it has some problems with fields of type `Option<T>` for some reason?

I don't see it as a requirement.

Alexendoo replied:
The actual visibility is what the lint is about though

> Fields prefixed with an `_` are inferred as unused, which suggests it should not be marked as `pub`, because marking it as `pub` infers it will be used.

`pub(crate)`, or `pub` where the field is not publicly reachable does not mark the field as used

What's the issue with `Option` fields?

blyxyas replied:
Hmm, now I'm not sure.
We could lint for any field that is public and its name starts with `_`, which is what the name of lint implies. Or we could analyze the description a little bit more and just lint when the `pub` is actually useful.

ParkMyCar commented:
> Just some polishing and I think this is ready! Thanks for the patience! ❤️ (=^･ω･^=)

Sweet! Made all of your suggested changes!

blyxyas commented:
Could you please rebase and squash these commits into 1?

ParkMyCar commented:
> Could you please rebase and squash these commits into 1 or 2 relevant ones?

Done!

blyxyas commented:
Parker, we have another problem (=◑ᆺ◐=)
I couldn't decide if we should take Alex's approach of just linting effective visibilities or all instances of `pub _*: T`, so I opened [this poll] and the public has spoken.

The final vote is that it should be up to the user via a configuration option. You can see in `lib.rs` how other lints use configuration options and mimic that. Sorry for the extra work, it's just what *the council desires* :sparkles: :judge:.

ParkMyCar commented:
> Parker, we have another problem (=◑ᆺ◐=) I couldn't decide if we should take Alex's approach of just linting effective visibilities or all instances of `pub _*: T`, so I opened [this poll](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/.60pub_underscore_fields.60.20lint.20cases) and the public has spoken.
> 
> The final vote is that it should be up to the user via a configuration option. You can see in `lib.rs` how other lints use configuration options and mimic that. Sorry for the extra work, it's just what _the council desires_ ✨ 🧑‍⚖️.

No worries! I would rather a helpful lint than an annoying one 😊 Thanks for setting up the poll! It might be a day or two until I can get back to this, but I'll update it based on the council's recommendation 🫡

bors commented:
:umbrella: The latest upstream changes (presumably #11791) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ParkMyCar commented:
@blyxyas @Alexendoo finally got a chance to come back to this, sorry for the delay!

I added a new config option with two variants, `PublicallyExported` (default) and `AllPubFields`. Like described in the Zulip thread, the `PublicallyExported` option only lints on public fields that are exported, whereas `AllPubFields` lints on whether or not the field is marked `pub`. I also updated the UI tests to cover both config options. Let me know if there are any other changes that should be made, thanks for iterating on this with me!

blyxyas commented:
Currently recuperating from surgery

r? Alexendoo

ParkMyCar commented:
@blyxyas definitely no rush, I hope recovery goes well :)

blyxyas commented on clippy_lints/src/pub_underscore_fields.rs:
````diff
@@ -0,0 +1,88 @@
+    /// }
+    /// ```
+    ///
+    /// // OR
````
```suggestion
    /// OR
```

ParkMyCar replied:
Fixed!

blyxyas commented on clippy_lints/src/pub_underscore_fields.rs:
```diff
@@ -0,0 +1,88 @@
+}
+
+fn start_with_pub(cx: &LateContext<'_>, span: Span) -> bool {
+    snippet_opt(cx, span)
+        .map(|s| s.as_str().starts_with("pub"))
+        .unwrap_or(false)
```
Don't need to look for `pub`, there's no visibility span if the visibility is not `pub`

```suggestion
    !span.is_empty()
```

ParkMyCar replied:
Ah ha good call, updated!

blyxyas commented on tests/ui-toml/pub_underscore_fields/pub_underscore_fields.rs:
NIT: Probably better to split this into `pub_underscore_fields/exported/pub_underscore_fields.rs` and `pub_underscore_fields/all_pub_fields/pub_underscore_fields.rs`, or find a way to make the `.stderr` to those directories, having those `pub_underscore_fields.*.stderr` seems uncomfortable.

ParkMyCar replied:
I originally used [revisions](https://rustc-dev-guide.rust-lang.org/tests/ui.html#cfg-revisions) because it's what other lints did, e.g. [absolute_paths](https://github.com/rust-lang/rust-clippy/tree/master/tests/ui-toml/absolute_paths). More than happy to use the multiple test files approach if you'd still prefer that :)

blyxyas replied:
Don't worry, they are all in the same parent directory, it should be understandable enough. (Even more if it's what other lints are doing, although I haven't seen it :thinking: )

blyxyas reviewed (changes requested):
Honestly, I'm very sorry that this PR is getting so long with a conversation so convoluted! Other PRs get resolved within 2 weeks, it's kinda shameful that this one is getting 10 months old.

Also, it's really impressive that you haven't lost interest! It should be merge-ready with just these few little changes though (´; ω ;｀)

blyxyas commented:
Also, I totally forgot that this isn't one of Alexendoo's PRs, I actually self-assigned this one to me :sweat_smile:. Let's finish this!

r? blyxyas
(Let's remove this one off Alexendoo's queue hehe :3)

ParkMyCar commented:
> Honestly, I'm very sorry that this PR is getting so long with a conversation so convoluted! Other PRs get resolved within 2 weeks, it's kinda shameful that this one is getting 10 months old.

No problem at all! This is an open source project, things generally move slow and that's okay :)

> Also, it's really impressive that you haven't lost interest! It should be merge-ready with just these few little changes though (´; ω ;｀)

Thank you!

blyxyas reviewed (approved):
LGTM! Thanks :heart: and sorry for the long delay. I'll approve it and merge it!

blyxyas commented:
@bors r+

bors commented:
:pushpin: Commit ea82130bac0e6d27763272695f4e329f89fa6eea has been approved by `blyxyas`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=blyxyas ea82130bac0e6d27763272695f4e329f89fa6eea -->
<!-- homu: {"type":"Approved","sha":"ea82130bac0e6d27763272695f4e329f89fa6eea","approver":"blyxyas","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

blyxyas commented:
@bors r-

bors commented:
:hourglass: Testing commit ea82130bac0e6d27763272695f4e329f89fa6eea with merge e93063421e683da707bc97f0b061863cd608b0b9...
<!-- homu: {"type":"BuildStarted","head_sha":"ea82130bac0e6d27763272695f4e329f89fa6eea","merge_sha":"e93063421e683da707bc97f0b061863cd608b0b9"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/7338052815/job/19980155475)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/actions/runs/7338052815/job/19980155475","builder_name":"checks-action_test"} -->

blyxyas commented on clippy_lints/src/pub_underscore_fields.rs:
````diff
@@ -0,0 +1,83 @@
+    ///     pub descriptor: usize,
+    /// }
+    /// ```
+    #[clippy::version = "1.76.0"]
````
```suggestion
    #[clippy::version = "1.77.0"]
```

... And update this...

ParkMyCar replied:
Done!

blyxyas reviewed (changes requested):
And also, could you run `cargo collect-metadata`

blyxyas reviewed (changes requested):
And also, could you run `cargo collect-metadata`

ParkMyCar commented:
> And also, could you run `cargo collect-metadata`

Updated to 1.77, ran `cargo collect-metadata`, and squashed all of the commits into one!

blyxyas commented on book/src/lint_configuration.md:
```diff
@@ -805,3 +805,14 @@ for _ in &mut *rmvec {}
+---
+**Affected lints:**
+* [`pub_underscore_fields`](https://rust-lang.github.io/rust-clippy/master/index.html#pub_underscore_fields)
+* [`or`](https://rust-lang.github.io/rust-clippy/master/index.html#or)
```
I'm not really sure why this is here

ParkMyCar replied:
Me either! It's what `cargo collect-metadata` generated, other lints seem to have similar "Affected lints" sections :)

blyxyas replied:
I meant the `or` lint, that doesn't exist xD

blyxyas commented:
@bors try

bors commented:
:hourglass: Trying commit 73a3636a23ad80058f4eb7bff6a76638a545e4a5 with merge 493b5a31558ca205f5fe5c3a13bca6652baf16bc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"73a3636a23ad80058f4eb7bff6a76638a545e4a5","merge_sha":"493b5a31558ca205f5fe5c3a13bca6652baf16bc"} -->

bors commented:
:sunny: Try build successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/7355404595/job/20023974710), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/7355404593/job/20023977446), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/7355404594/job/20024103351)
Build commit: 493b5a31558ca205f5fe5c3a13bca6652baf16bc (`493b5a31558ca205f5fe5c3a13bca6652baf16bc`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7355404594/job/20024103351","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7355404595/job/20023974710","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7355404593/job/20023977446"},"merge_sha":"493b5a31558ca205f5fe5c3a13bca6652baf16bc"} -->

blyxyas commented:
I changed that `,` to a `;`, this is a bug in our metadata collector, it seems that it just gets any word after a comma. I already pushed this to your fork.

Let's hope that everything goes well this time (=^･^=)

blyxyas commented:
@bors r+

bors commented:
:pushpin: Commit fa7dd1c4e0106ce0346236801653188689dbaa43 has been approved by `blyxyas`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=blyxyas fa7dd1c4e0106ce0346236801653188689dbaa43 -->
<!-- homu: {"type":"Approved","sha":"fa7dd1c4e0106ce0346236801653188689dbaa43","approver":"blyxyas","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit fa7dd1c4e0106ce0346236801653188689dbaa43 with merge 174a0d7be6f18bf6470a46d107e4b3f0b870c315...
<!-- homu: {"type":"BuildStarted","head_sha":"fa7dd1c4e0106ce0346236801653188689dbaa43","merge_sha":"174a0d7be6f18bf6470a46d107e4b3f0b870c315"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/7355996476/job/20025468441), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/7355996471/job/20025470585), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/7355996473/job/20025555311)
Approved by: blyxyas
Pushing 174a0d7be6f18bf6470a46d107e4b3f0b870c315 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"blyxyas","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7355996473/job/20025555311","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7355996476/job/20025468441","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7355996471/job/20025470585"},"merge_sha":"174a0d7be6f18bf6470a46d107e4b3f0b870c315"} -->

ParkMyCar commented:
Woo! Thank you @blyxyas for reviewing this PR and getting it merged, I appreciate it! 😊

GrantGryczan commented:
It's spelled "publicly", not "publically". I recommend fixing that in the `PublicallyExported` value. (If needed, I can submit a separate issue for this.)

blyxyas commented:
I'll fix this myself. Thanks for the report.
